pyomo/repn/plugins/parameterized_standard_form.py

```diff
@@ -0,0 +1,233 @@
+        row_index_ptr = self.indptr
+        nrows = self.shape[0]
+
+        num_nonzeros = len(csr_data)
+        csc_data = np.empty(csr_data.shape[0], dtype=object)
+        row_index = np.empty(num_nonzeros, dtype=int)
+        # tally the nonzeros in each column
+        col_index_ptr = np.zeros(self.shape[1], dtype=int)
+        for i in col_index:
+            col_index_ptr[int(i)] += 1
+
+        # cumulative sum the tally to get the column index pointer
+        cum_sum = 0
+        for i, tally in enumerate(col_index_ptr):
+            col_index_ptr[i] = cum_sum
+            cum_sum += tally
+        # we leave off the last entry (the number of nonzeros) because we are
+        # going to do the cute scipy thing and it will get 'added' at the end
+        # when we shift right (by which I mean it will conveniently already be
+        # there)
+
+        # Now we are actually going to mess up what we just did while we
+        # construct the row index, but it's beautiful because "mess up" just
+        # means we increment everything by one, so we can fix it at the end.
+        for row in range(nrows):
+            for j in range(row_index_ptr[row], row_index_ptr[row + 1]):
+                col = col_index[j]
+                dest = col_index_ptr[col]
+                row_index[dest] = row
+                # Note that the data changes order because now we are looking
+                # for nonzeros through the columns rather than through the rows.
+                csc_data[dest] = csr_data[j]
+
+                col_index_ptr[col] += 1
+
+        # fix the column index pointer by inserting 0 at the beginning--the rest
+        # is right because we know each entry got incremented by 1 in the loop
+        # above.
+        col_index_ptr = np.insert(col_index_ptr, 0, 0)
+
+        return _CSCMatrix((csc_data, row_index, col_index_ptr), self.shape)
+
+    def todense(self):
+        nrows = self.shape[0]
+        col_index = self.indices
+        row_index_ptr = self.indptr
+        data = self.data
+
+        dense = np.zeros(self.shape, dtype=object)
+
+        for row in range(nrows):
+            for j in range(row_index_ptr[row], row_index_ptr[row + 1]):
+                dense[row, col_index[j]] = data[j]
+
+        return dense
+
+
+class _CSCMatrix(_SparseMatrixBase):
+    def todense(self):
+        ncols = self.shape[1]
+        row_index = self.indices
+        col_index_ptr = self.indptr
+        data = self.data
+
+        dense = np.zeros(self.shape, dtype=object)
+
+        for col in range(ncols):
+            for j in range(col_index_ptr[col], col_index_ptr[col + 1]):
+                dense[row_index[j], col] = data[j]
+
+        return dense
+
+    def sum_duplicates(self):
+        ncols = self.shape[1]
+        row_index = self.indices
+        col_index_ptr = self.indptr
+        data = self.data
+
+        num_non_zeros = 0
+        col_end = 0
+        for i in range(ncols):
+            jj = col_end
```

Comment:
I agree that this is terrible, but I'm reimplementing C code from scipy, so stuck with their variable names for now so that looking between the two is easy... Not sure if that's a very good excuse though. :P

Reply:
We should add a comment indicating the inspiration (source) of this function.